conformance/packages/conformance-tests/src/name_server/rfc5155.rs

```diff
@@ -0,0 +1,376 @@
+use std::collections::BTreeMap;
+use std::net::Ipv4Addr;
+
+use dns_test::client::{Client, DigSettings, DigStatus};
+use dns_test::name_server::NameServer;
+use dns_test::record::{Record, RecordType, NSEC3};
+use dns_test::{Network, Result, FQDN};
+
+///  An NSEC3 RR is said to "match" a name if the owner name of the NSEC3 RR is the same as the
+///  hashed owner name of that name.
+fn find_match<'a>(name_hash: &str, records: &'a BTreeMap<String, NSEC3>) -> Option<&'a NSEC3> {
+    records.get(name_hash)
+}
+
+/// An NSEC3 RR is said to cover a name if the hash of the name or "next closer" name falls between
+/// the owner name and the next hashed owner name of the NSEC3.  In other words, if it proves the
+/// nonexistence of the name, either directly or by proving the nonexistence of an ancestor of the
+/// name.
+fn find_cover<'a>(name_hash: &str, records: &'a BTreeMap<String, NSEC3>) -> Option<&'a NSEC3> {
+    let (hash, candidate) = records
+        // Find the greater hash that is less or equal than the name's hash.
+        .range(..=name_hash.to_owned())
+        .last()
+        // If no value is less or equal than the name's hash, it means that the name's hash is out
+        // of range and the last record covers it.
+        .or_else(|| records.last_key_value())?;
+
+    // If the found hash is exactly the name's hash, return None as it wouldn't be proving its
+    // nonexistence. Otherwise return the RR with that hash.
+    (hash != name_hash).then_some(candidate)
+}
+
+/// This proof consists of (up to) two different NSEC3 RRs:
+/// - An NSEC3 RR that matches the closest (provable) encloser.
+/// - An NSEC3 RR that covers the "next closer" name to the closest encloser.
+fn closest_encloser_proof<'a>(
+    closest_encloser_hash: &str,
+    next_closer_name_hash: &str,
+    records: &'a BTreeMap<String, NSEC3>,
```

Comment:
it may improve readability to create a newtype over `BTreeMap<String, NSEC3>` and have these 3 functions (find_match, find_cover, closest_encloser_proof) be methods on that type. and the constructor of that newtype can probably take `&SignedZoneFile` as input

Reply:
Good idea :) Done